internal/db/storage.go

```diff
@@ -0,0 +1,247 @@
+func (d *DBClient) BeginTx(ctx context.Context) (context.Context, TxInterface, error) {
+	tx, err := d.db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelReadCommitted, ReadOnly: false})
+	if err != nil {
+		return ctx, nil, err
+	}
+
+	return ContextWithTx(ctx, tx), tx, nil
+}
+
+// ContextWithTx returns a new context with the transaction attached.
+func ContextWithTx(ctx context.Context, tx TxInterface) context.Context {
+	return context.WithValue(ctx, txContextKey{}, tx)
+}
+
+// TxFromContext extracts a transaction from the context, returning nil if none exists.
+func TxFromContext(ctx context.Context) TxInterface {
+	if tx, ok := ctx.Value(txContextKey{}).(TxInterface); ok {
+		return tx
+	}
+	return nil
+}
+
+// lazyTxFromContext extracts a lazy transaction holder from the context.
+func lazyTxFromContext(ctx context.Context) *lazyTx {
+	if lt, ok := ctx.Value(lazyTxContextKey{}).(*lazyTx); ok {
+		return lt
+	}
+	return nil
+}
+
+// contextWithLazyTx returns a new context with a lazy transaction holder attached.
+func contextWithLazyTx(ctx context.Context, lt *lazyTx) context.Context {
+	return context.WithValue(ctx, lazyTxContextKey{}, lt)
+}
+
+// WithTx executes a function within a transaction context.
+// The transaction is created lazily on first database access.
+// If the function returns an error, the transaction is rolled back.
+// Otherwise, the transaction is committed.
+// If no database operations occurred, no transaction is created or committed.
+func (d *DBClient) WithTx(ctx context.Context, fn func(context.Context) error) error {
+	lt := &lazyTx{
+		db:     d.db,
+		logger: d.logger,
+	}
+	txCtx := contextWithLazyTx(ctx, lt)
+
+	defer func() {
+		// Only rollback if transaction was started and not committed
+		if lt.isStarted() && !lt.committed {
+			if err := lt.tx.Rollback(); err != nil && err != sql.ErrTxDone {
+				d.logger.Errorf("failed to rollback transaction: %v", err)
+			}
+		}
+	}()
+
+	if err := fn(txCtx); err != nil {
+		return err
+	}
+
+	// Only commit if transaction was actually started
+	if lt.isStarted() {
+		if err := lt.tx.Commit(); err != nil {
+			return fmt.Errorf("failed to commit transaction: %v", err)
+		}
+		lt.committed = true
+	}
+
+	return nil
+}
+
+func (d *DBClient) Close() {
+	if d.db != nil {
+		_ = d.db.Close()
+	}
+
+	if d.pool != nil {
+		d.pool.Close()
+	}
+}
+
+// NewDBClient creates a new DBClient instance with the provided DSN and configuration options.
+func NewDBClient(dsn string, tracingEnabled bool, tracer tracing.TracingInterface, monitor monitoring.MonitorInterface, logger logging.LoggerInterface) *DBClient {
+	config, err := pgxpool.ParseConfig(dsn)
+	if err != nil {
+		logger.Fatalf("DSN validation failed, shutting down, err: %v", err)
+	}
+
+	if tracingEnabled {
+		// otelpgx.NewTracer will use default global TracerProvider, just like our tracer struct
+		config.ConnConfig.Tracer = otelpgx.NewTracer()
+	}
+
+	pool, err := pgxpool.NewWithConfig(context.Background(), config)
```

Comment:
don't forget to set the idle timeout on the connection (if possible) to prevent transaction rollback issues to bork the connection and make it unusable